io.github.TTimo.GtkRadiant.json

```diff
@@ -0,0 +1,220 @@
+{
+  "app-id": "io.github.TTimo.GtkRadiant",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "3.28",
```

Comment:
Any reason you're using so old runtime?

Reply:
The current source doesn't build out of the box against newer SDKs. A bit of investigation suggests it would take some non trivial refactoring and developer time to modernize.

GtkRadiant was written 20+ years ago against the Gtk 2 libraries. It uses an obsolete module (gtkglext) that depends on more obsolete pieces (pangox).

It has been increasingly difficult to build GtkR on modern distributions; we are looking to leverage flatpak's older SDKs and Platforms to help with this problem.

Reply:
> A bit of investigation suggests it would take some non trivial refactoring and developer time to modernize.

You mean just a few patches?

Reply:
> > A bit of investigation suggests it would take some non trivial refactoring and developer time to modernize.
> 
> You mean just a few patches?

@TTimo is a bit humble but he is the best people on earth to evaluate the amount of work required to implement this or that. He was actually part of the id Software team (the company that initially developed this software and the games this software is meant to produce content for) and was in charge of the Linux ports of those games and applications. I assume he is the one who actually ported Quake 3 to Linux, the one who ported this application from Windows MFC to GTK back in the days, ported this application to more recent games like Quake Live, etc. After that he worked for Valve to port games to Linux ([example of interview on the topic](https://www.gamingonlinux.com/2016/09/interview-with-timothee-besset-formerly-of-id-software-who-helped-port-rocket-league-to-linux/)). TTimo is likely doing the maintenance of this software since two decades. So not only he knows very well about porting software to this  or that technology, but he also knows very well how to evaluate the cost of such effort in both term of time, effort or money. If he says his investigation suggests it would take some non trivial refactoring and developer time to modernize, he fully means it and this is not underrated in any way.

Now, GtkRadiant is a fully volunteer project, including TTimo's maintenance effort. The summed available volunteer time of all GtkRadiant contributors is very small (maybe less than a couple of hours per year). This software is still the reference on the topic, though.

Let's elaborate on this sentence:

> GtkRadiant was written 20+ years ago against the Gtk 2 libraries. It uses an obsolete module (gtkglext) that depends on more obsolete pieces (pangox).

I don't know exactly what port to what new version of what libraries requires the newer SDKs, but let's just pick an obvious example: GTK.

What if someone comes and ask “why you would not port it to GTK3, it's just a few patch”? Here is the answer:

This software uses GTK 2 with GtkGLExt with OpenGL 1. GTK3 replaced GtkGLExt with GtkGLArea that don't support OpenGL 1. Rewriting the renderer from OpenGL 1 to OpenGL core (what supports GTK3's GtkGlArea) would require a similar amount of work that would be required to port a game from OpenGL to Vulkan (to make a comparison with things from today). This is huge. As a comparison, the [Dæmon Engine](https://github.com/DaemonEngine/Daemon) who has an OpenGL core reimplementation of the OpenGL 1 Quake 3 renderer haven't reached  100% compatibility after 17 years of volunteer work. Luckily the GtkRadiant renderer only implement a very small subset of those features, but really, doing such port would not be easy. Instead of a few patch it would be rewriting some core components from scratch.

Hopefully GTK3 GtkGLArea has a undocumented feature to run OpenGL 1, but this undocumented feature only exists on Linux and Windows, not on macOS. So to keep GtkRadiant buildable on both Linux, Windows and macOS, GtkRadiant would be able to be built with both GTK2 and GTK3 with the same source tree. As an example, [NetRadiant](https://netradiant.gitlab.io/) implemented a wrapper over both GTK2 and GTK3 to allow this, and after years of work of volunteer time, the GTK3 implementation is still not fully working and can still not share an OpenGL context across windows, meaning the application has to run every OpenGL view in a single window instead of allowing to use multiple window (eventually on multiple screens) and other bugs (power consumption is much higher because it still renders the frame multiple time per second instead of only redrawing it when something changes). That means for NetRadiant, the GTK3 build was never released and is only available for those building it themselves. After months or years of work starting from today to do something similar, GtkRadiant would be in the same half-baked situation.

I don't know what is required to use a more recent runtime, but if more recent runtime requires GTK3, a port may require several months of full time work of a seasoned developer.

Other changes in other components may also require huge rewrite in similar ways.

GtkRadiant hasn't seen a Linux build since 2013 while there are more recent versions for Windows and macOS, one of the reason seems to be that it is hard to rely on old libraries on Linux. You'll also notice this flatpak work started5 years ago, that also gives a glance on how few volunteer time is available to make this software move forward.

Since flatpak can host multiple applications running multiple runtimes from various ages, I can understand that flatpak appears as a good candidate for running applications using obsolete technologies. If flathub refuses to include applications using obsolete technologies just on principle of refusing applications using old technologies despite being technically compatible with that, flathub would be no better than other packages managers. Such limitation being a technical limitation (like traditional package repositories have) or a moral principle would have the exact same effect.

Reply:
Yeah, I think it's a valid point to make. If the software is old, and the only sensible way to make it work is to use an old runtime, I'd say it's acceptable.

Reply:
I see a Gtk2 folder in the shared-modules repo. We're not going to upgrade to Gtk 3, but we may still be able to move to a newer Sdk reasonably easily if we leverage the shared-modules backports.

Reply:
it starts too, but then this is a workstation.

Reply:
oh and this is using freedesktop sdk 22.08

Reply:
It's all there https://github.com/hfiguiere/flathub/tree/gtkradiant

Reply:
Thanks for the edits! We're on the new runtime now.